### PR TITLE
Observaciones de estructura

### PR DESCRIPTION
**Pull request by rogs-lv**
1. Agrega meaningful
2. Correccion de complejidad para SumArray
3. Modificaciones a PR
4. Código limpio

### CODE REVIEW

File: Algorithms.Test/ShuffleTheArrayTest.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        /// Constructor
+        /// </summary>
+        public ShuffleTheArrayTest() {
+            _shuffleTheArray    = new ShuffleTheArray();
+        }
+
+        [Fact]
+        public void TestCase1() {
+            //Input
+            int[] numsArray     = new int[] { 2, 5, 1, 3, 4, 7 };
```
Comment (jorgetinoco) [inline Algorithms.Test/ShuffleTheArrayTest.cs]: adding the spaces between the assignation is generaly not a good practice, since adding a new line requires adjustment on each line

File: Algorithms/RunningSumArray.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        /// <param name="nums">Array de enteros</param>
+        /// <returns>Array de enteros</returns>
+        public int[] RunningSum(int[] nums)
+        {
+            int len = nums.Length;
+            for (int i = 0; i < len; i++)
+            {
+                if (i > 0)
+                    nums[i] = nums[i - 1] + nums[i];
+                else
```
Comment (jorgetinoco) [inline Algorithms/RunningSumArray.cs]: there's a way you can remove this if/else statement.

File: Algorithms.Test/ShuffleTheArrayTest.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            //Input
+            int[] numsArray     = new int[] { 1, 1, 2, 2 };
+            int elements        = 2;
+            //Expected Output
+            int[] outputArray   = new int[] { 1, 2, 1, 2 };
+            var result      = _shuffleTheArray.Shuffle(numsArray, elements);
+            Assert.Equal(outputArray, result);
+        }
+        [Fact]
+        public void TestCase4() {
```
Comment (jorgetinoco) [inline Algorithms.Test/ShuffleTheArrayTest.cs]: Imagine this is a real project. TestCase4 doesn't tell me what this test actually do. You can check unit tests best practices here:
https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests

File: Algorithms.Test/ShuffleTheArrayTest.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            int elements        = 420;
```
Comment (jorgetinoco) [inline Algorithms.Test/ShuffleTheArrayTest.cs]: As you can see, all your tests are the happy path. 
What happens if I sent `elements = -1`, what happens if I sent `nums Array = new int[0]` and then `elements = 10`... etc etc. That's the meaning of edge cases

File: Algorithms/Program.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Algorithms
+{
+    public class Program
```
Comment (jorgetinoco) [inline Algorithms/Program.cs]: probably you don't need this class

File: Algorithms/RunningSumArray.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                    nums[i] = nums[i];
+            }
+            return nums;
+        }
+        /// <summary>
+        /// Big O 
+        /// Complejidad: O(N2) - Se tienen dos ciclos recorriendo el mismo arreglo, y el segundo lo recorre
```
Comment (jorgetinoco) [inline Algorithms/RunningSumArray.cs]: 👍

File: README.md [hunk 7]
```diff
@@ -5,6 +5,16 @@ Leetcode exercise for training. A solution with multiple console applications.
 
 Run `dotnet build` to build the project.
 
+## Test
```
Comment (jorgetinoco) [inline README.md:8]: A good practice in the README is to mention the minimum version you need to run the project. For example:

## Build

To build this project you need at least NetCore 3.1.
Then you can build the project with

```bash
dotnet build
```

File: Algorithms/RunningSumArray.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+namespace Algorithms
+{
+    public class RunningSumArray
+    {
+        /// <summary>
+        /// Big O
+        /// Complejidad: O(n) - El arreglo solo se recorre una unica vez y se le resta la posición 
+        /// anterior que ya tiene el ultimo valor acumulado
+        /// </summary>
+        /// <param name="nums">Array de enteros</param>
```
Comment (jorgetinoco) [inline Algorithms/RunningSumArray.cs]: Good coding practices:
If the parameter is called `nums` it can be anything, and then you end up with documentation like this one `Array de enteros`, here are two things:
- Imagine your code is international, it is always a good practice to do your code in a "neutral" language as english so everyone can understand it
- What could be a better variable naming? `numArray` for example and that could give you more information without reading the documentation.


Apply the same to other parts of the code, for example in shuffleArray, what's `l`, `j` ?

**Review by jorgetinoco — CHANGES_REQUESTED** [review]
Hey @rogs-lv I did some comments.

Some of them may look "overkill" for this project but remember the reason behind this exercise, which is improve in all aspects, that's why I'm being so picky about clean code, tests etc.

Also feel free to add here the exercise you sent me, the palindrome solution

**Comment by jorgetinoco** [comment]
@rogs-lv 
Also try to follow git/github good practices for this small project.
One example I have right now is about commits, it is easier to follow the history of the commits and review changes if commits are separated meaningfully. For example in this case I only see 1 commit `Agrega observaciones de estructura` which in my head I can only think: `which observations?` `which structure?`.
Eventhough you added the comment in the commit, it would be easier to see the following (example) commit structure:

```bash
fix variables to have meaningful names
fix SumArray, make it O(n)
Improve clean code
```
Inside each commit you can explain more about what it is about, but then if I want to review only how you changed SumArray to be O(n) I can go to that commit and see the changes.

https://medium.com/@nawarpianist/git-commit-best-practices-dab8d722de99

File: Algorithms.Test/ShuffleTheArrayTest.cs (hunk 1 above)
Comment (rogs-lv) [reply Algorithms.Test/ShuffleTheArrayTest.cs]: Got it!. I read once, that my code should be easy to read, like a book. I assumed if I used tabs or extra space. It would achieve the goal. I think ... i didn't understand well hehe

File: Algorithms/RunningSumArray.cs (hunk 2 above)
Comment (rogs-lv) [reply Algorithms/RunningSumArray.cs]: Yes, There's a way. Actually first element of array never is affected with new value. Therefore loop for can start in 1.

File: Algorithms/Program.cs (hunk 5 above)
Comment (rogs-lv) [reply Algorithms/Program.cs]: Got it!.

File: Algorithms/RunningSumArray.cs (hunk 6 above)
Comment (rogs-lv) [reply Algorithms/RunningSumArray.cs]: 👍 Cool!

File: README.md (hunk 7 above)
Comment (rogs-lv) [reply README.md:8]: Perfect! :)

File: Algorithms.Test/ShuffleTheArrayTest.cs (hunk 1 above)
Comment (jorgetinoco) [reply Algorithms.Test/ShuffleTheArrayTest.cs]: That's true, it should be easy to read as a book. One part of that is to use spaces, but in a different way for example:
What's easier to read, this:
```js
let y = 10;
let x = 1;
if(y==='division')
return x/y;
else if(y==='mult')
return x*y;
return x+y;
```

or this:
```js
let operator = 10;
let myNumber = 1;

if (operation === 'division')
   return myNumber / operator;
else if (operation === 'mult')
  return myNumber * operator;

return myNumber + operator;
```

If you see, I changed 2 things, one was spaces/tabs and the second was variable naming. This "easy to read" heavily applies to variable naming and "cleaningless" of the code, reading x + y doesn't give you any context or what x or y means, while a different variable name can give you more information so you can keep reading the code without reading documentation, reading external functions or any other thing


So yes, that's true adding spaces helps, but you should think also in maintainability, for example

```js
let x          = true
let y          = false
let z          = 1 
```

and what happen if you want to add a variable called `superLongName`:
```js
let x          = true
let y          = false
let z          = 1 
let superLongName = true
```

Then you have either of 2 problems, they are not aligned anymore, so not easy to read, or you need to update 3 lines instead of just adding your variable.

File: Algorithms.Test/ShuffleTheArrayTest.cs (hunk 3 above)
Comment (rogs-lv) [reply Algorithms.Test/ShuffleTheArrayTest.cs]: I check the information 💯

File: Algorithms.Test/ShuffleTheArrayTest.cs (hunk 4 above)
Comment (rogs-lv) [reply Algorithms.Test/ShuffleTheArrayTest.cs]: Got it. I'll change my unit tests

File: Algorithms/RunningSumArray.cs (hunk 8 above)
Comment (rogs-lv) [reply Algorithms/RunningSumArray.cs]: OK, So i'll change comments.

File: Algorithms.Test/ShuffleTheArrayTest.cs (hunk 1 above)
Comment (rogs-lv) [reply Algorithms.Test/ShuffleTheArrayTest.cs]: Great!! I definitely didn't get it right. Now i see the problem 👏

File: Algorithms/ShuffleTheArray.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            if (numsArray == null) return new int[] { };
+            if (numsArray.Length == 0) return numsArray;
+            if (n < 1 || n > 500 || ((n * 2) != numsArray.Length)) return new int[] { };
+
+            int lenArray = numsArray.Length;
+            int[] newArr = new int[lenArray];
+            bool flag = true;
+            int j = n;
+            int l = 0;
```
Comment (jorgetinoco) [inline Algorithms/ShuffleTheArray.cs]: What about this? What's `l`?

File: Algorithms/ShuffleTheArray.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        /// <returns>Returns the array with combined values</returns>
+        public int[] Shuffle(int[] numsArray, int n)
+        {
+            if (numsArray == null) return new int[] { };
+            if (numsArray.Length == 0) return numsArray;
+            if (n < 1 || n > 500 || ((n * 2) != numsArray.Length)) return new int[] { };
+
+            int lenArray = numsArray.Length;
+            int[] newArr = new int[lenArray];
+            bool flag = true;
```
Comment (jorgetinoco) [inline Algorithms/ShuffleTheArray.cs]: can you think in a better name than `flag`?

File: Algorithms/ShuffleTheArray.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            int j = n;
+            int l = 0;
+            for (int i = 0; i < lenArray; i++)
+            {
+                if (flag)
+                {
+                    newArr[i] = numsArray[l];
+                    l++;
+                    flag = false;
```
Comment (jorgetinoco) [inline Algorithms/ShuffleTheArray.cs]: Can you think in a way to improve this `flag` code? In how many place you need to update the flag logic if for some reason it changes?

File: Algorithms.Test/RunningSumArrayTest.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        [Fact]
+        public void RunningSum_ArrayNull_ReturnArraryEmpty() {
+            int[] numsArray = null;
+
+            var result = _runningSumArray.RunningSum(numsArray);
+            Assert.Empty(result);
+        }
+        [Fact]
+        public void RunningSum_ArrayWithOneElement_ReturnSameArray()
```
Comment (jorgetinoco) [inline Algorithms.Test/RunningSumArrayTest.cs:39]: better tests!

File: Algorithms.Test/RunningSumArrayTest.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+namespace Algorithms.Test
+{
+    public class RunningSumArrayTest
+    {
+        RunningSumArray _runningSumArray;
+        public RunningSumArrayTest() {
+            _runningSumArray = new RunningSumArray();
+        }
+        [Fact]
+        public void RunningSum_SmallArray_ReturnRunningSumOfNumbers()
```
Comment (jorgetinoco) [inline Algorithms.Test/RunningSumArrayTest.cs:15]: What about testing negative numbers inside the array?

File: Algorithms/RunningSumArray.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+{
+    public class RunningSumArray
+    {
+        /// <summary>
+        /// Big O
+        /// Complexity: O(n) - The array is only traversed once, and the new value is the current position plus the value of the previous position, with accumulated value
+        /// </summary>
+        /// <param name="numsArray">Numbers array</param>
+        /// <returns>Return the running sum of numbers.</returns>
+        public int[] RunningSum(int[] numsArray)
```
Comment (jorgetinoco) [inline Algorithms/RunningSumArray.cs:16]: what do you think about this code now? Comparing it to the first version

**Review by jorgetinoco — CHANGES_REQUESTED** [review]

File: Algorithms/RunningSumArray.cs (hunk 14 above)
Comment (rogs-lv) [reply Algorithms/RunningSumArray.cs:16]: I think it's better, because by use the same array to overwrite value, i don't need more memory space. I also think it's cleaner to reduce the number of lines of code.

File: Algorithms/ShuffleTheArray.cs (hunk 10 above)
Comment (rogs-lv) [reply Algorithms/ShuffleTheArray.cs]: Sure, i think so.

File: Algorithms/ShuffleTheArray.cs (hunk 11 above)
Comment (rogs-lv) [reply Algorithms/ShuffleTheArray.cs]: Yes, i can. I think is better if i change logic of method.

**Review by rogs-lv — COMMENTED** [review]
Applied

File: Algorithms.Test/RunningSumArrayTest.cs [hunk 15]
```diff
@@ -40,6 +40,21 @@ public void RunningSum_ArrayWithOneElement_ReturnSameArray()
         {
             int[] numsArray = new int[] { 3 };
 
+            var result = _runningSumArray.RunningSum(numsArray);
+            Assert.Equal(numsArray, result);
+        }
+        [Fact]
+        public void RunningSum_ArrayWithNegativeNumber_ReturnSameArray() {
+            int[] numsArray = new int[] { -1000001, 2, 1 };
```
Comment (jorgetinoco) [inline Algorithms.Test/RunningSumArrayTest.cs:48]: nice catch! 
But my question remains... what would happen if I send { -1, -2, -3, 5 }
(I know the code, so I know it would work, but there're missing tests about those cases)

**Review by jorgetinoco — CHANGES_REQUESTED** [review]

File: Algorithms/ShuffleTheArray.cs [hunk 16]
```diff
@@ -7,38 +7,59 @@ namespace Algorithms
...
-                    l++;
-                    flag = false;
-                }
-                else
-                {
-                    newArr[i] = numsArray[j];
-                    j++;
-                    flag = true;
-                }
+                newArray[i + i] = numsArray[i];
```
Comment (jorgetinoco) [inline Algorithms/ShuffleTheArray.cs:24]: 🚀

File: Algorithms/ShuffleTheArray.cs (hunk 10 above)
Comment (jorgetinoco) [reply Algorithms/ShuffleTheArray.cs]: Just want to point this out. After thinking in better naming, you figure it out how to clean your code and ended up with a better version of the solution. That's what you need to do everytime :)

Comment (rogs-lv) [reply Algorithms/ShuffleTheArray.cs]: I understand, thanks for this advice 👌  👍 . In fact I always try, but sometimes I get stuck 😕

**Comment by rogs-lv** [comment]
Add a new exercise with two solutions, the first solution had a little problem and was slower. But after thinking about what you told me about the name of the variables 🤔. I saw a new way to change the algorithm 🤓 . 
In this exercise i applied all the things we have been working on. 😃

File: Algorithms/PalindromeNumber.cs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        /// <summary>
+        /// Complexity: O(N) - We use only one for, because the variant is to determine the midpoint
+        /// </summary>
+        /// <param name="number">value to evaluate</param>
+        /// <returns>true if is palindrome or false if is not</returns>
+        public bool IsPalindrome(int number) {
+            if (number < 0) return false; //Numbers negatives never will true
+            if (number <= 9 && number >= 0) return true; // Numbers with only digits always will true
+
+            int leng = number.ToString().Length;
```
Comment (jorgetinoco) [inline Algorithms/PalindromeNumber.cs]: length of what?

File: Algorithms/PalindromeNumber.cs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+            char[] arr = number.ToString().ToCharArray();
```
Comment (jorgetinoco) [inline Algorithms/PalindromeNumber.cs]: what is arr?

File: Algorithms/PalindromeNumber.cs [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+            int midpoint = (leng%2) == 1 ? (leng - 1) / 2 : leng / 2; // 1 is not even number || 0 is even number
+            for (int i = 0; i < midpoint; i++)
```
Comment (jorgetinoco) [inline Algorithms/PalindromeNumber.cs:23]: good one! This is the best way to check for palindromes. 
You will see different things like reverse, or things like that but this way is "better", and actually this part could be O(n/2) since you are not iterating the whole array, you are iterating half of it

File: Algorithms.Test/PalindromeNumberTest.cs [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        [Fact]
+        public void IsPalindrome_NumberPositive_ReturnTrue() {
+            int number = 121;
+
+            var result = _palindromeNumber.IsPalindrome(number);
+            Assert.True(result);
+        }
+        [Fact]
+        public void IsPalindrome_NumberNegative_ReturnFalse()
```
Comment (jorgetinoco) [inline Algorithms.Test/PalindromeNumberTest.cs:43]: I'm guessing you used the one in leetcode for this exercise.
No need to do it, but how would you do it if for the problem didn't matter if it is negative or positive?
for example -515 should be true. What do you think you could do?

File: Algorithms/ValidParentheses.cs [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+					if (charsOfClose[characters.Peek()] == stringCharacters[i])
+						characters.Pop();
+					else
+						return false;
+				}
+				else
+					return false;
+				i++;
+			}
+			return characters.Count == 0 ? true : false;
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs]: how could you improve this last line?

File: Algorithms/ValidParentheses.cs [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+		/// <summary>
+		/// Big O
+		/// Complexity: O(n) : we go through the array of characters given and depending on the length of the string is the time it takes to push and remove the character from the stack 
+		/// </summary>
+		/// <param name="stringCharacters">string of characters</param>
+		/// <returns>true if all characters was closed, false if some characters wasn't closed</returns>
+		public bool IsValid(string stringCharacters) {
+			if (stringCharacters[0] == ')' || stringCharacters[0] == '}' || stringCharacters[0] == ']') return false;
+
+			Stack<char> characters = new Stack<char>();
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs:18]: nice

File: Algorithms/ValidParentheses.cs [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+			Dictionary<char, char> charsOfClose = new Dictionary<char, char>();
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs:19]: nice! good way of improving the complexity :)

File: Algorithms/ValidParentheses.cs [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+			charsOfClose.Add('(', ')');
+			charsOfClose.Add('{', '}');
+			charsOfClose.Add('[', ']');
+			int lengthString = stringCharacters.Length, i = 0;
+
+			while (i < lengthString)
+			{
+				if (stringCharacters[i] == '(' || stringCharacters[i] == '{' || stringCharacters[i] == '[')
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs]: Check this documentation: https://github.com/ryanmcdermott/clean-code-javascript#encapsulate-conditionals
It is for JS, but the concepts can be applied in all languages (almost everything at least)
Check specific the one I shared `Encapsulate conditionals`

File: Algorithms/ValidParentheses.cs [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+    public class ValidParentheses
+    {
+		/// <summary>
+		/// Big O
+		/// Complexity: O(n) : we go through the array of characters given and depending on the length of the string is the time it takes to push and remove the character from the stack 
+		/// </summary>
+		/// <param name="stringCharacters">string of characters</param>
+		/// <returns>true if all characters was closed, false if some characters wasn't closed</returns>
+		public bool IsValid(string stringCharacters) {
+			if (stringCharacters[0] == ')' || stringCharacters[0] == '}' || stringCharacters[0] == ']') return false;
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs]: what happens with an empty or null string?
You missed that unit test :)

File: Algorithms/JewelsAndStones.cs [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+        public int NumJewelsInStones(string charsToSearch, string charsString)
+        {
+            if (charsToSearch == null || charsString == null) return 0;
+
+            Dictionary<char, int> dictionaryJewels = new Dictionary<char, int>();
+            int sum = 0;
+            for (int i = 0; i < charsToSearch.Length; i++)
+            {
+                dictionaryJewels[charsToSearch[i]] = 0;
+            }
```
Comment (jorgetinoco) [inline Algorithms/JewelsAndStones.cs:26]: What could be a "break first" approach here?
What if you don't have any charsToSearch? do you still need to check the string?

File: Algorithms/JewelsAndStones.cs [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+
+            for (int j = 0; j < charsString.Length; j++)
+            {
+                if (dictionaryJewels.ContainsKey(charsString[j]))
+                    dictionaryJewels[charsString[j]]++;
+            }
+
+            foreach (var i in dictionaryJewels)
```
Comment (jorgetinoco) [inline Algorithms/JewelsAndStones.cs:34]: is there a way you can remove this last foreach?

File: Algorithms/ValidParentheses.cs [hunk 28]
```diff
@@ -22,22 +22,38 @@ public class ValidParentheses
...
+		/// <returns>return true or false</returns>
+		private bool validOpenChars(string stringCharacters, int i) { 
+			return (stringCharacters[i] == '(' || stringCharacters[i] == '{' || stringCharacters[i] == '[') ? true : false;
+		}
+		/// <summary>
+		/// Validate start of string with closing characters and string is empty
+		/// </summary>
+		/// <param name="stringCharacters">string of characters</param>
+		/// <returns>return true or false</returns>
+		private bool validCloseOrEmptyChars(string stringCharacters) {
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs]: there are a couple of things you could improve here:
the function name + parameter suggests it will compare characters (in plural) to close or empty chars.
however the function only validates the starting character (position 0).
If you see your 2 functions validOpenChars and validCloseChars, the name looks similar but they do different things.

So my suggestion would be change the function name to something like `isFirstCharCloseOrEmpty` or do something similar than with validOpenChars, including the position

Comment (jorgetinoco) [reply Algorithms/ValidParentheses.cs]: The other thing, is that this line is redundant:
return (boolean statement) ? true : false;

you can always `return boolean statement`

File: Algorithms/ValidParentheses.cs [hunk 29]
```diff
@@ -22,22 +22,38 @@ public class ValidParentheses
...
-				else
+				} else
 					return false;
 				i++;
 			}
-			return characters.Count == 0 ? true : false;
+			return characters.Count == 0 & true;
+		}
+		/// <summary>
+		/// Valid open chars
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs]: If you are going to include comments, make them more useful than just the name of the function

File: Algorithms/ValidParentheses.cs [hunk 30]
```diff
@@ -22,22 +22,38 @@ public class ValidParentheses
...
+		/// <param name="i">index</param>
+		/// <returns>return true or false</returns>
+		private bool validOpenChars(string stringCharacters, int i) { 
+			return (stringCharacters[i] == '(' || stringCharacters[i] == '{' || stringCharacters[i] == '[') ? true : false;
+		}
+		/// <summary>
+		/// Validate start of string with closing characters and string is empty
+		/// </summary>
+		/// <param name="stringCharacters">string of characters</param>
+		/// <returns>return true or false</returns>
```
Comment (jorgetinoco) [inline Algorithms/ValidParentheses.cs]: when does it returns true or false? what needs to happen to get each?

**Comment by jorgetinoco** [comment]
@rogs-lv just checking in if you had time to review the latest comments 😄

**Comment by rogs-lv** [comment]
It's ready 👍  💯  😃
